marconi/src/Marconi/Index/MintBurn.hs

```diff
@@ -0,0 +1,310 @@
+    | Just to' <- slotMaybe to -> ("slotNo <= :toSlot" , [":toSlot" := to'])
+    | otherwise -> ("FALSE", [])
+    where
+      slotMaybe :: C.ChainPoint -> Maybe C.SlotNo
+      slotMaybe = \case
+        C.ChainPoint slotNo _ -> Just slotNo
+        C.ChainPointAtGenesis -> Nothing
+
+groupBySlotNo :: [TxMintEvent] -> [TxMintEvent]
+groupBySlotNo events = events
+  & sort
+  & groupBy (\e1 e2 -> txMintEventSlotNo e1 == txMintEventSlotNo e2 && txMintEventBlockHeaderHash e1 == txMintEventBlockHeaderHash e2)
+  & concatMap (\case e : es -> [ TxMintEvent (txMintEventSlotNo e) (txMintEventBlockHeaderHash e) $ txMintEventTxAssets =<< (e :| es) ]
+                     _      -> [])
+
+-- * Indexer
+
+data MintBurnHandle = MintBurnHandle
+  { sqlConnection :: SQL.Connection
+  , securityParam :: Int
+  }
+
+type MintBurnIndexer = RI.State MintBurnHandle
+
+type instance RI.StorablePoint MintBurnHandle = C.ChainPoint
+
+type instance RI.StorableMonad MintBurnHandle = IO
+
+newtype instance RI.StorableEvent MintBurnHandle
+  = MintBurnEvent TxMintEvent
+
+data instance RI.StorableQuery MintBurnHandle
+  = ByTxIdAndPolicyId C.TxId C.PolicyId
+  | Everything
+
+data instance RI.StorableResult MintBurnHandle
+  = MintBurnResult [Row]
+
+instance RI.Queryable MintBurnHandle where
+  queryStorage queryInterval memoryEvents (MintBurnHandle sqlCon _k) query = case query of
+    Everything                      -> toResult <$> sqliteSelectAll sqlCon interval
+    ByTxIdAndPolicyId txId policyId -> toResult <$> sqliteSelectByTxIdPolicyId sqlCon interval txId policyId
+    where
+      toResult storedEvents = MintBurnResult $ do
+        TxMintEvent slotNo blockHeaderHash txAssets <- storedEvents <> map coerce (toList memoryEvents)
+        (txId, mintAssets) <- NE.toList txAssets
+        MintAsset policyId assetName quantity redeemerIx redeemerData <- NE.toList mintAssets
+        pure (slotNo, blockHeaderHash, txId, policyId, assetName, quantity, redeemerIx, redeemerData)
+
+      interval = intervalToWhereClause queryInterval
+
+instance RI.HasPoint (RI.StorableEvent MintBurnHandle) C.ChainPoint where
+  getPoint (MintBurnEvent e) = C.ChainPoint (txMintEventSlotNo e) (txMintEventBlockHeaderHash e)
+
+instance RI.Buffered MintBurnHandle where
+  persistToStorage events h@(MintBurnHandle sqlCon _k) = do
+    sqliteInsert sqlCon (map coerce $ toList events)
+    pure h
+
+  getStoredEvents (MintBurnHandle sqlCon k) =
+    map MintBurnEvent . fromRows <$> SQL.query sqlCon query (SQL.Only k)
+    where
+      query =
+        " SELECT slotNo, blockHeaderHash, txId, policyId, assetName, quantity  \
+        \        redeemerIx, redeemerData                                      \
+        \   FROM minting_policy_events                                         \
+        \  WHERE slotNo >= (SELECT MAX(slotNo) - ? FROM minting_policy_events) \
```

Comment:
I would replace this one with a `LIMIT k`. It will give more events which for resume (where this is used) is a good thing.

Reply:
Hmm, right. I wonder if `getStoredEvents` is a good function to use for resume points anyway? This idea I took from the ScriptTx indexer as you used it there. But it might be better to write the sql directly into `resumeFromStorage` instead?

Trying to understand the meaning of `getStoredEvents`: it's used [here in RewindableIndex.Storable.getEvents](https://github.com/input-output-hk/plutus-apps/blob/main/rewindable-index/src/RewindableIndex/Storable.hs#L246) -- does it need to get *all* events, or some of the events? In ScriptTx indexer it seems to get only some of the events, while for `RewindableIndex.Storable.getEvents` it appears to get all of them?